neurips23/streaming/base_postgres.py

```diff
@@ -0,0 +1,262 @@
+                    try:
+                        cur.execute(f"SELECT id FROM test_tbl ORDER BY vec_col {self.query_op} %s LIMIT {k}", (query_vec, ))
+                    except Exception as e:
+                        raise Exception(f"Error '{e}' when querying with k={k}\nQuery vector was:\n{query_vec}") from e
+
+                    result_tuples = cur.fetchall()
+
+                    result_ids = list(map(lambda tup: tup[0], result_tuples))
+
+                    if len(result_ids) < k:
+                        # Pad with -1 if we have less than k results. This is only needed if the
+                        # index-access method cannot guarantee returning k results.
+                        #
+                        # As of today, this is only possible with PostgresPgvectorHnsw when
+                        # ef_search < k.
+                        result_ids.extend([-1] * (k - len(result_ids)))
+
+                    batch_result_id_lists.append(result_ids)
+
+            return batch_result_id_lists
+
+        total_queries = len(X)
+
+        result_id_lists = []
+
+        # Run the batch_query function in parallel
+        with concurrent.futures.ThreadPoolExecutor(max_workers=self.n_query_conns) as executor:
+            chunk_size = ceil(total_queries / self.n_query_conns)
+            query_futures = []
+            for conn_idx, query_vec_start_idx in enumerate(range(0, total_queries, chunk_size)):
+                query_vec_end_idx = min(query_vec_start_idx + chunk_size, total_queries)
+                query_futures.append(executor.submit(batch_query, conn_idx, query_vec_start_idx, query_vec_end_idx))
+
+            start_time = time.time()
+
+            # wait for all futures to complete
+            done, not_done = concurrent.futures.wait(query_futures)
+
+            exec_time = time.time() - start_time
+
+            log_dict = {
+                'query': total_queries,
+                'exec_time': exec_time
+            }
+
+            print('Timing:', log_dict)
+
+            assert len(not_done) == 0
+            assert len(done) == len(query_futures)
+
+            # retrieve the results in the order they were submitted to avoid messing up the order
+            for query_future in query_futures:
+                batch_result_id_lists = query_future.result()
+                result_id_lists.extend(batch_result_id_lists)
+
+        self.res = np.vstack(result_id_lists, dtype=np.int32)
+
+    def set_query_arguments(self, query_args):
+        # close any existing connections
+        for conn in self.conns:
+            conn.close()
+
+        # By using a temporary connection, truncate the table since set_query_arguments() is called
+        # before each testing phase with new set of query params.
+        with psycopg.connect(PG_CONN_STR, autocommit=True) as conn:
+            with conn.cursor() as cur:
+                cursor_print_and_execute(cur, "TRUNCATE test_tbl")
```

Comment:
Although the other algo implementations don't seem to be doing this, I thought that we have to reset all the data in main table and the index here, i.e., before switching to a different query-args set. Does that make sense?